### PR TITLE
[backup] `db-restore auto`

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation

This mode automatically decides which backups to use to achieve a target version. The restore command line gets dramatically simpler in most cases.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

Played on local machine.

``` text
➜  libra git:(restore-coordinator) ✗ cargo run -p backup-cli --bin db-restore -- --target-db-dir ~/libradb_restored_5 --target-version 3000 auto local-fs --dir ~/libra_backup
   Compiling backup-cli v0.1.0 (/Users/aldenhu/work/libra/storage/backup/backup-cli)
    Finished dev [unoptimized + debuginfo] target(s) in 16.55s
     Running `target/debug/db-restore --target-db-dir /Users/aldenhu/libradb_restored_5 --target-version 3000 auto local-fs --dir /Users/aldenhu/libra_backup`
Warning: Can't find transaction backup that contains the target version, will restore as much as possible
Epoch ending restore started. Manifest: /Users/aldenhu/libra_backup/epoch_ending_0-/epoch_ending.manifest
Finished restoring epoch ending info.
Epoch ending restore succeeded.
State snapshot restore started. Manifest: /Users/aldenhu/libra_backup/state_ver_2600/state.manifest
Finished restoring state snapshot.
State snapshot restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_0-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_200-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_400-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_600-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_800-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_1000-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_1200-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_1400-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_1600-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_1800-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_2000-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_2200-/transaction.manifest
Transaction restore succeeded.
Transaction restore started. Manifest: /Users/aldenhu/libra_backup/transaction_2400-/transaction.manifest
Transaction restore succeeded.
Restore finished.
```

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs [hunk 1]
```diff
@@ -43,6 +43,20 @@ impl EpochEndingRestoreController {
...
+        );
+        self.run_impl()
+            .await
+            .map_err(|e| anyhow!("Epoch ending restore failed: {}", e))?;
+        println!("Epoch ending restore succeeded.");
+        Ok(())
+    }
+}
+
+impl EpochEndingRestoreController {
```
Comment (wqfish) [inline storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs:58]: does this need to be a separate `impl` block?

File: storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs [hunk 2]
```diff
@@ -49,6 +49,20 @@ impl StateSnapshotRestoreController {
...
+        );
+        self.run_impl()
+            .await
+            .map_err(|e| anyhow!("State snapshot restore failed: {}", e))?;
+        println!("State snapshot restore succeeded.");
+        Ok(())
+    }
+}
+
+impl StateSnapshotRestoreController {
```
Comment (wqfish) [inline storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs:64]: same here and below

File: storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs (hunk 1 above)
Comment (msmouse) [reply storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs:58]: Most structs in this crate (whereever I remembered), I extracted the pub interfaces (usually few) to a separate impl...

**Pull request closed without merging**